File: rust/src/core/readiness.rs
//! Readiness primitives - one monotonic deadline, checks that report evidence.
//!
//! Navigation used to answer "did the page settle?" and then throw the answer
//! away: `wait_for_url_stabilization` returned `false` on timeout, the caller
//! discarded it, and the navigation was reported as a success anyway. These
//! types make the answer impossible to discard - a readiness wait carries the
//! evidence for every check it ran, and one deadline bounds all of them.

use serde_json::{json, Value};
use std::fmt;
use std::time::{Duration, Instant};

/// How a readiness wait ended.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ReadinessStatus {
    /// Every check was satisfied.
    #[default]
    Ready,
    /// The budget ran out before every check could answer.
    TimedOut,
    /// A check reported, within budget, that it was not satisfied.
    Failed,
}

impl fmt::Display for ReadinessStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ReadinessStatus::Ready => write!(f, "ready"),
            ReadinessStatus::TimedOut => write!(f, "timed_out"),
            ReadinessStatus::Failed => write!(f, "failed"),
        }
    }
}

/// A monotonic budget shared by every check in one readiness wait.
///
/// The clock is monotonic on purpose: a wall-clock jump - an NTP correction, a
/// suspended laptop - must not turn a five second budget into a five minute
/// one. `remaining` never grows, so no caller can be handed a larger budget
/// than the one it asked for.
#[derive(Debug, Clone)]
pub struct Deadline {
    started_at: Instant,
    timeout: Duration,
}

impl Deadline {
    /// Start a deadline with the given total budget.
    #[must_use]
    pub fn new(timeout: Duration) -> Self {
        Self {
            started_at: Instant::now(),
            timeout,
        }
    }

    /// Total budget this deadline was created with.
    #[must_use]
    pub fn timeout(&self) -> Duration {
        self.timeout
    }

    /// Time spent so far.
    #[must_use]
    pub fn elapsed(&self) -> Duration {
        self.started_at.elapsed()
    }

    /// Elapsed milliseconds, for evidence records.
    #[must_use]
    pub fn elapsed_ms(&self) -> u128 {
        self.elapsed().as_millis()
    }

    /// Budget left, saturating at zero.
    #[must_use]
    pub fn remaining(&self) -> Duration {
        self.timeout.saturating_sub(self.elapsed())
    }

    /// Whether the budget is spent.
    #[must_use]
    pub fn expired(&self) -> bool {
        self.remaining().is_zero()
    }

    /// Sleep for at most the remaining budget.
    pub async fn sleep_at_most(&self, requested: Duration) {
        let capped = requested.min(self.remaining());
        if !capped.is_zero() {
            tokio::time::sleep(capped).await;
        }
    }
}

/// Evidence for one check that ran.
#[derive(Debug, Clone, PartialEq)]
pub struct CheckRecord {
    /// Check name.
    pub name: String,
    /// Whether the check was satisfied.
    pub satisfied: bool,
    /// Whether the check could not run and was skipped.
    pub skipped: bool,
    /// Milliseconds into the wait at which the check started.
    pub started_at_ms: u128,
    /// Milliseconds the check itself took.
    pub elapsed_ms: u128,
    /// Structured detail describing what the check observed.
    pub detail: Value,
}

impl CheckRecord {
    /// Record a satisfied check.
    #[must_use]
    pub fn satisfied(name: impl Into<String>, detail: Value) -> Self {
        Self {
            name: name.into(),
            satisfied: true,
            skipped: false,
            started_at_ms: 0,
            elapsed_ms: 0,
            detail,
        }
    }

    /// Record a check that was not satisfied.
    #[must_use]
    pub fn unsatisfied(name: impl Into<String>, detail: Value) -> Self {
        Self {
            name: name.into(),
            satisfied: false,
            skipped: false,
            started_at_ms: 0,
            elapsed_ms: 0,
            detail,
        }
    }

    /// Record a check that could not run.
    #[must_use]
    pub fn skipped(name: impl Into<String>, reason: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            satisfied: false,
            skipped: true,
            started_at_ms: 0,
            elapsed_ms: 0,
            detail: json!({ "reason": reason.into() }),
        }
    }

    /// Attach the timing window the check occupied.
    #[must_use]
    pub fn with_timing(mut self, started_at_ms: u128, elapsed_ms: u128) -> Self {
        self.started_at_ms = started_at_ms;
        self.elapsed_ms = elapsed_ms;
        self
    }
}

/// The outcome of a readiness wait, with the evidence behind it.
#[derive(Debug, Clone, Default)]
pub struct ReadinessOutcome {
    /// How the wait ended.
    pub status: ReadinessStatus,
    /// Whether every check was satisfied.
    pub ready: bool,
    /// Names of satisfied checks.
    pub satisfied: Vec<String>,
    /// Names of checks that reported "not satisfied".
    pub failed: Vec<String>,
    /// Names of checks that could not run.
    pub skipped: Vec<String>,
    /// Names of checks that never started because the budget ran out.
    pub pending: Vec<String>,
    /// Evidence for every check that ran.
    pub evidence: Vec<CheckRecord>,
    /// Total milliseconds spent.
    pub elapsed_ms: u128,
    /// The budget the wait was given.
    pub timeout_ms: u128,
}

impl ReadinessOutcome {
    /// Start an empty outcome bound to a deadline's budget.
    #[must_use]
    pub fn new(deadline: &Deadline) -> Self {
        Self {
            status: ReadinessStatus::Ready,
            ready: true,
            timeout_ms: deadline.timeout().as_millis(),
            ..Self::default()
        }
    }

    /// Fold one check's evidence into the outcome.
    pub fn record(&mut self, record: CheckRecord) {
        if record.skipped {
            self.skipped.push(record.name.clone());
        } else if record.satisfied {
            self.satisfied.push(record.name.clone());
        } else {
            self.failed.push(record.name.clone());
        }
        self.evidence.push(record);
    }

    /// Note a check that never got to run.
    pub fn defer(&mut self, name: impl Into<String>) {
        self.pending.push(name.into());
    }

    /// Settle the status against the deadline once every check has reported.
    #[must_use]
    pub fn finish(mut self, deadline: &Deadline) -> Self {
        self.ready = self.failed.is_empty() && self.pending.is_empty();
        self.elapsed_ms = deadline.elapsed_ms();
        self.status = if self.ready {
            ReadinessStatus::Ready
        } else if deadline.expired() {
            ReadinessStatus::TimedOut
        } else {
            ReadinessStatus::Failed
        };
        self
    }

    /// A one-line summary suitable for a result reason.
    #[must_use]
    pub fn summary(&self) -> String {
        if self.ready {
            return format!("page ready after {}ms", self.elapsed_ms);
        }
        format!(
            "page not ready ({}) after {}ms; failed=[{}] pending=[{}]",
            self.status,
            self.elapsed_ms,
            self.failed.join(", "),
            self.pending.join(", ")
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_deadline_never_hands_out_more_budget_than_it_was_given() {
        // Regression guard for issue #89: the JS implementation computed
        // `max(60000, timeout - elapsed)`, so a 10ms budget became 60s.
        let deadline = Deadline::new(Duration::from_millis(10));

        assert!(deadline.remaining() <= Duration::from_millis(10));
        assert_eq!(deadline.timeout(), Duration::from_millis(10));
    }

    #[test]
    fn remaining_saturates_at_zero() {
        let deadline = Deadline::new(Duration::ZERO);

        assert!(deadline.expired());
        assert_eq!(deadline.remaining(), Duration::ZERO);
    }

    #[tokio::test]
    async fn sleeping_is_capped_by_the_remaining_budget() {
        let deadline = Deadline::new(Duration::from_millis(20));

        deadline.sleep_at_most(Duration::from_secs(30)).await;

        assert!(deadline.elapsed() < Duration::from_secs(1));
    }

    #[test]
    fn an_unsatisfied_check_never_reports_ready() {
        // Regression guard for issue #89: "did not stabilize" used to be
        // discarded and the operation reported as a success.
        let deadline = Deadline::new(Duration::from_secs(5));
        let mut outcome = ReadinessOutcome::new(&deadline);

        outcome.record(CheckRecord::unsatisfied(
            "url_stable_for",
            json!({ "url": "https://example.com/" }),
        ));
        let outcome = outcome.finish(&deadline);

        assert!(!outcome.ready);
        assert_eq!(outcome.status, ReadinessStatus::Failed);
        assert_eq!(outcome.failed, vec!["url_stable_for".to_string()]);
        assert!(outcome.summary().contains("url_stable_for"));
    }

    #[test]
    fn an_expired_budget_reports_timed_out() {
        let deadline = Deadline::new(Duration::ZERO);
        let mut outcome = ReadinessOutcome::new(&deadline);

        outcome.record(CheckRecord::unsatisfied("url_stable_for", json!({})));
        let outcome = outcome.finish(&deadline);

        assert_eq!(outcome.status, ReadinessStatus::TimedOut);
    }

    #[test]
    fn a_skipped_check_does_not_block_readiness() {
        let deadline = Deadline::new(Duration::from_secs(5));
        let mut outcome = ReadinessOutcome::new(&deadline);

        outcome.record(CheckRecord::skipped("network_idle_for", "no tracker"));
        let outcome = outcome.finish(&deadline);

        assert!(outcome.ready);
        assert_eq!(outcome.skipped, vec!["network_idle_for".to_string()]);
    }

    #[test]
    fn checks_that_never_ran_are_pending_not_satisfied() {
        let deadline = Deadline::new(Duration::from_secs(5));
        let mut outcome = ReadinessOutcome::new(&deadline);

        outcome.record(CheckRecord::satisfied("url_stable_for", json!({})));
        outcome.defer("verify_navigation");
        let outcome = outcome.finish(&deadline);

        assert!(!outcome.ready);
        assert_eq!(outcome.pending, vec!["verify_navigation".to_string()]);
    }

    #[test]
    fn evidence_carries_the_timing_window_of_each_check() {
        let record = CheckRecord::satisfied("url_stable_for", json!({})).with_timing(5, 12);

        assert_eq!(record.started_at_ms, 5);
        assert_eq!(record.elapsed_ms, 12);
    }

    #[test]
    fn statuses_render_the_documented_names() {
        assert_eq!(ReadinessStatus::Ready.to_string(), "ready");
        assert_eq!(ReadinessStatus::TimedOut.to_string(), "timed_out");
        assert_eq!(ReadinessStatus::Failed.to_string(), "failed");
    }
}

File: js/src/core/readiness.js
/**
 * Readiness primitives - one monotonic deadline, composable checks, evidence.
 *
 * `waitForPageReady` used to answer "is the page ready?" with a hard-coded
 * sequence of waits, a floor that could outlive the caller's timeout, and a
 * constant `true`. This module supplies the pieces that make the answer
 * truthful: a deadline that only ever hands out non-negative remaining budget,
 * checks that report what they observed, and a result that distinguishes
 * "ready" from "we ran out of time".
 */

/** Statuses a readiness wait can end in. */
export const READINESS_STATUS = Object.freeze({
  READY: 'ready',
  TIMED_OUT: 'timed_out',
  FAILED: 'failed',
});

/**
 * Request classes that never go idle on their own. Waiting for them is waiting
 * for the timeout, so network idle ignores them unless a caller opts back in.
 */
export const LONG_LIVED_REQUEST_PATTERNS = Object.freeze([
  /^wss?:\/\//iu,
  /\/(?:socket\.io|sockjs|websocket|ws)(?:[/?]|$)/iu,
  /\/(?:event-?stream|sse|stream)(?:[/?]|$)/iu,
  /(?:^|\.)(?:google-analytics|googletagmanager|doubleclick|segment|mixpanel|amplitude|hotjar|sentry|datadoghq|newrelic)\./iu,
  /\/(?:analytics|telemetry|beacon|collect|metrics|heartbeat|ping|poll|longpoll)(?:[/?]|$)/iu,
]);

/**
 * True when a URL belongs to a request class that is expected to stay open.
 *
 * @param {string} url - Request URL
 * @param {RegExp[]} [patterns] - Patterns to test, defaults to the built-in list
 * @returns {boolean} Whether the request should be ignored by idle checks
 */
export function isLongLivedRequest(
  url,
  patterns = LONG_LIVED_REQUEST_PATTERNS
) {
  if (typeof url !== 'string' || url.length === 0) {
    return false;
  }
  return patterns.some((pattern) => pattern.test(url));
}

/**
 * Create a monotonic deadline shared by every check in one readiness wait.
 *
 * The clock is monotonic on purpose: a wall-clock jump (NTP correction, a
 * suspended laptop) must not turn a five second budget into a five minute one.
 *
 * @param {Object} options - Configuration options
 * @param {number} options.timeout - Total budget in milliseconds
 * @param {Function} [options.now] - Monotonic clock, defaults to performance.now
 * @returns {{startedAt: number, timeoutMs: number, elapsedMs: Function, remainingMs: Function, expired: Function}} Deadline handle
 */
export function createDeadline(options = {}) {
  const { timeout, now = () => performance.now() } = options;

  if (!Number.isFinite(timeout) || timeout < 0) {
    throw new TypeError('createDeadline requires a non-negative timeout');
  }

  const startedAt = now();
  const elapsed = () => Math.max(0, now() - startedAt);
  // `expired` is derived from `remainingMs` rather than measured separately:
  // rounding otherwise lets a caller read a remaining budget of zero while the
  // deadline still calls itself live, and a check that stopped because its
  // budget was gone would then be reported as failed instead of timed out.
  const remainingMs = () => Math.max(0, Math.round(timeout - elapsed()));

  return {
    startedAt,
    timeoutMs: timeout,
    elapsedMs: () => Math.round(elapsed()),
    remainingMs,
    expired: () => remainingMs() === 0,
  };
}

/**
 * Sleep for at most the deadline's remaining budget.
 *
 * @param {number} ms - Requested delay
 * @param {Object} deadline - Deadline from {@link createDeadline}
 * @returns {Promise<void>} Resolves after the capped delay
 */
export async function sleepWithinDeadline(ms, deadline) {
  const capped = Math.min(ms, deadline.remainingMs());
  if (capped <= 0) {
    return;
  }
  await new Promise((resolve) => setTimeout(resolve, capped));
}

/** Marker resolved by the expiry timer in {@link runWithinDeadline}. */
const DEADLINE_REACHED = Symbol('deadline-reached');

/**
 * Run an operation under a deadline, reporting expiry instead of waiting.
 *
 * Engine probes carry timeouts of their own - Playwright's locator default is
 * 30 seconds - so an operation given a 3 second budget could spend ten times
 * that waiting for an element a navigation had already taken away. The budget
 * the caller asked for has to win.
 *
 * @param {Object} deadline - Deadline from {@link createDeadline}, or null for no bound
 * @param {Function} run - Zero-argument function returning a promise
 * @returns {Promise<{timedOut: boolean, value: *}>} Outcome, or expiry
 */
export async function runWithinDeadline(deadline, run) {
  if (!deadline) {
    return { timedOut: false, value: await run() };
  }

  const remaining = deadline.remainingMs();
  if (remaining <= 0) {
    return { timedOut: true, value: undefined };
  }

  const operation = run();
  // An abandoned operation still settles on its own schedule. Handle its
  // rejection up front so it cannot surface as an unhandled one after the
  // deadline has already won the race.
  Promise.resolve(operation).catch(() => {});

  let timer;
  const expiry = new Promise((resolve) => {
    timer = setTimeout(() => resolve(DEADLINE_REACHED), remaining);
  });

  try {
    const outcome = await Promise.race([operation, expiry]);
    if (outcome === DEADLINE_REACHED) {
      return { timedOut: true, value: undefined };
    }
    return { timedOut: false, value: outcome };
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Resolve the engine adapter a check needs, building it lazily when the caller
 * only supplied a factory.
 *
 * @param {Object} context - Check context
 * @returns {Promise<Object>} Engine adapter
 */
async function resolveAdapter(context) {
  const adapter = context.adapter ?? (await context.getAdapter?.());
  if (!adapter) {
    throw new Error('readiness check requires an engine adapter');
  }
  return adapter;
}

function normalizeSampleResult(value) {
  if (typeof value === 'boolean') {
    return { stable: value, detail: {} };
  }
  return { stable: Boolean(value?.stable), detail: value?.detail ?? {} };
}

/**
 * Build a check that requires a sampler to report stability for a period.
 *
 * @param {Object} options - Configuration options
 * @param {string} options.name - Check name reported in evidence
 * @param {Function} options.sample - Async sampler returning boolean or {stable, detail}
 * @param {number} [options.stableForMs=500] - How long the sampler must stay stable
 * @param {number} [options.intervalMs=100] - Sampling interval
 * @param {number} [options.consecutiveSamples=1] - Extra consecutive stable samples required
 * @returns {{name: string, run: Function}} Readiness check
 */
export function stableCheck(options = {}) {
  const {
    name,
    sample,
    stableForMs = 500,
    intervalMs = 100,
    consecutiveSamples = 1,
  } = options;

  if (!name || typeof sample !== 'function') {
    throw new TypeError('stableCheck requires a name and a sample function');
  }

  return {
    name,
    async run(context) {
      const { deadline } = context;
      let stableSince = null;
      let streak = 0;
      let lastDetail = {};

      while (!deadline.expired()) {
        const { stable, detail } = normalizeSampleResult(await sample(context));
        lastDetail = detail;

        if (stable) {
          streak += 1;
          stableSince ??= deadline.elapsedMs();
          const heldFor = deadline.elapsedMs() - stableSince;
          if (heldFor >= stableForMs && streak >= consecutiveSamples) {
            return {
              satisfied: true,
              detail: { ...lastDetail, heldForMs: heldFor },
            };
          }
        } else {
          stableSince = null;
          streak = 0;
        }

        if (deadline.remainingMs() === 0) {
          break;
        }
        await sleepWithinDeadline(intervalMs, deadline);
      }

      return {
        satisfied: false,
        detail: { ...lastDetail, reason: 'deadline reached' },
      };
    },
  };
}

/**
 * Require the page URL to stop changing for a period.
 *
 * @param {Object} [options] - Configuration options
 * @param {number} [options.stableForMs=1000] - Required quiet period
 * @param {number} [options.intervalMs=200] - Sampling interval
 * @param {number} [options.consecutiveSamples=1] - Extra consecutive stable samples
 * @returns {{name: string, run: Function}} Readiness check
 */
export function urlStableFor(options = {}) {
  const {
    stableForMs = 1000,
    intervalMs = 200,
    consecutiveSamples = 1,
  } = options;
  let lastUrl;

  return stableCheck({
    name: 'urlStableFor',
    stableForMs,
    intervalMs,
    consecutiveSamples,
    sample: ({ page, onUrlSample }) => {
      const url = page.url();
      onUrlSample?.(url);
      const stable = lastUrl === undefined ? false : url === lastUrl;
      lastUrl = url;
      return { stable, detail: { url } };
    },
  });
}

/**
 * Require the network tracker to report idle within the remaining budget.
 *
 * @param {Object} [options] - Configuration options
 * @param {number} [options.idleForMs] - Idle window, defaults to the tracker's own
 * @returns {{name: string, run: Function}} Readiness check
 */
export function networkIdleFor(options = {}) {
  const { idleForMs } = options;

  return {
    name: 'networkIdleFor',
    async run({ deadline, networkTracker }) {
      if (!networkTracker) {
        return { skipped: true, detail: { reason: 'no network tracker' } };
      }

      const timeout = deadline.remainingMs();
      if (timeout === 0) {
        return { satisfied: false, detail: { reason: 'no remaining budget' } };
      }

      const idle = await networkTracker.waitForNetworkIdle({
        timeout,
        ...(idleForMs === undefined ? {} : { idleTime: idleForMs }),
      });

      return {
        satisfied: Boolean(idle),
        detail: {
          pendingCount: networkTracker.getPendingCount?.() ?? null,
          pendingUrls: idle ? [] : (networkTracker.getPendingUrls?.() ?? []),
        },
      };
    },
  };
}

/**
 * Require the DOM to stop mutating for a period. Uses a cheap structural
 * fingerprint so it works identically on both engines and needs no injection.
 *
 * @param {Object} [options] - Configuration options
 * @param {number} [options.stableForMs=500] - Required quiet period
 * @param {number} [options.intervalMs=100] - Sampling interval
 * @param {number} [options.consecutiveSamples=2] - Extra consecutive stable samples
 * @returns {{name: string, run: Function}} Readiness check
 */
export function domStableFor(options = {}) {
  const {
    stableForMs = 500,
    intervalMs = 100,
    consecutiveSamples = 2,
  } = options;
  let lastFingerprint;

  return stableCheck({
    name: 'domStableFor',
    stableForMs,
    intervalMs,
    consecutiveSamples,
    sample: async (context) => {
      const adapter = await resolveAdapter(context);
      const fingerprint = await adapter.evaluateOnPage(() => {
        const { body } = document;
        return {
          nodes: document.getElementsByTagName('*').length,
          length: body ? body.innerHTML.length : 0,
          readyState: document.readyState,
        };
      });
      const key = `${fingerprint.nodes}:${fingerprint.length}`;
      const stable =
        lastFingerprint === key && fingerprint.readyState !== 'loading';
      lastFingerprint = key;
      return { stable, detail: fingerprint };
    },
  });
}

/**
 * Require every image currently in the viewport to have finished decoding.
 *
 * @param {Object} [options] - Configuration options
 * @param {number} [options.intervalMs=150] - Sampling interval
 * @returns {{name: string, run: Function}} Readiness check
 */
export function visibleImages(options = {}) {
  const { intervalMs = 150 } = options;

  return stableCheck({
    name: 'visibleImages',
    stableForMs: 0,
    intervalMs,
    consecutiveSamples: 1,
    sample: async (context) => {
      const adapter = await resolveAdapter(context);
      const state = await adapter.evaluateOnPage(() => {
        const images = [...document.images];
        const inViewport = images.filter((image) => {
          const box = image.getBoundingClientRect();
          return (
            box.bottom > 0 &&
            box.right > 0 &&
            box.top < window.innerHeight &&
            box.left < window.innerWidth
          );
        });
        return {
          total: inViewport.length,
          pending: inViewport.filter((image) => !image.complete).length,
        };
      });
      return { stable: state.pending === 0, detail: state };
    },
  });
}

/**
 * Wrap a caller-supplied predicate as a readiness check.
 *
 * @param {Object} options - Configuration options
 * @param {Function} options.fn - Predicate receiving the check context
 * @param {string} [options.name='predicate'] - Name reported in evidence
 * @param {number} [options.intervalMs=100] - Polling interval
 * @returns {{name: string, run: Function}} Readiness check
 */
export function predicate(options = {}) {
  const { fn, name = 'predicate', intervalMs = 100 } = options;

  if (typeof fn !== 'function') {
    throw new TypeError('predicate requires a function');
  }

  return stableCheck({
    name,
    stableForMs: 0,
    intervalMs,
    consecutiveSamples: 1,
    sample: async (context) => ({ stable: Boolean(await fn(context)) }),
  });
}

/**
 * Run readiness checks in order against a single deadline and report evidence.
 *
 * Checks run to completion even after one fails, so the result explains the
 * whole picture rather than only the first problem. Checks that never started
 * are reported as pending, which is what tells a caller the wait was cut short
 * rather than genuinely unsatisfied.
 *
 * @param {Object} options - Configuration options
 * @param {Array} options.checks - Readiness checks to run
 * @param {Object} options.deadline - Deadline from {@link createDeadline}
 * @param {Object} options.context - Context handed to every check
 * @returns {Promise<Object>} Structured readiness result
 */
export async function runReadinessChecks(options = {}) {
  const { checks = [], deadline, context = {} } = options;

  const satisfied = [];
  const failed = [];
  const skipped = [];
  const pending = [];
  const evidence = [];

  for (const [index, check] of checks.entries()) {
    if (deadline.expired() && failed.length > 0) {
      pending.push(...checks.slice(index).map((entry) => entry.name));
      break;
    }

    const startedAtMs = deadline.elapsedMs();
    let outcome;
    try {
      outcome = await check.run({ ...context, deadline });
    } catch (error) {
      outcome = { satisfied: false, detail: { error: error.message } };
    }

    const record = {
      name: check.name,
      satisfied: Boolean(outcome.satisfied),
      skipped: Boolean(outcome.skipped),
      startedAtMs,
      elapsedMs: deadline.elapsedMs() - startedAtMs,
      detail: outcome.detail ?? {},
    };
    evidence.push(record);

    if (record.skipped) {
      skipped.push(check.name);
    } else if (record.satisfied) {
      satisfied.push(check.name);
    } else {
      failed.push(check.name);
    }
  }

  const ready = failed.length === 0 && pending.length === 0;
  const timedOut = !ready && deadline.expired();

  return {
    status: ready
      ? READINESS_STATUS.READY
      : timedOut
        ? READINESS_STATUS.TIMED_OUT
        : READINESS_STATUS.FAILED,
    ready,
    checks: { satisfied, failed, skipped, pending },
    evidence,
    elapsedMs: deadline.elapsedMs(),
    timeoutMs: deadline.timeoutMs,
  };
}
